runtime/lua/vim/ui/img.lua

```diff
@@ -0,0 +1,557 @@
+end
+
+---Returns an iterator over the chunks of the image, returning the chunk, byte position, and
+---an indicator of whether the current chunk is the last chunk.
+---
+---If `base64=true`, will encode the bytes using base64 before iterating chunks.
+---Takes an optional size to indicate how big each chunk should be, defaulting to 4096.
+---
+---Examples:
+---
+---```lua
+----- Some predefined image
+---local img = vim.ui.img.new({ ... })
+---
+------@param chunk string
+------@param pos integer
+------@param last boolean
+---img:chunks():each(function(chunk, pos, last)
+---  vim.print("Chunk bytes", chunk)
+---  vim.print("Chunk starts at", pos)
+---  vim.print("Is last chunk", last)
+---end)
+---```
+---
+---@param opts? {base64?:boolean, size?:integer}
+---@return Iter
+function M:chunks(opts)
+  opts = opts or {}
+
+  -- Chunk size, defaulting to 4k
+  local chunk_size = opts.size or 4096
+
+  local bytes = self.bytes
+  if not bytes or bytes == '' then
+    return vim.iter(function()
+      return nil, nil, nil
+    end)
+  end
+
+  if opts.base64 then
+    bytes = vim.base64.encode(bytes)
+  end
+
+  local pos = 1
+  local len = string.len(bytes)
+
+  return vim.iter(function()
+    -- If we are past the last chunk, this iterator should terminate
+    if pos > len then
+      return nil, nil, nil
+    end
+
+    -- Get our next chunk from [pos, pos + chunk_size)
+    local end_pos = pos + chunk_size - 1
+    local chunk = bytes:sub(pos, end_pos)
+
+    -- If we have a chunk available, mark as such
+    local last = true
+    if string.len(chunk) > 0 then
+      last = not (end_pos + 1 <= len)
+    end
+
+    -- Mark where our current chunk is positioned
+    local chunk_pos = pos
+
+    -- Update our global position
+    pos = end_pos + 1
+
+    return chunk, chunk_pos, last
+  end)
+end
+
+---@class (exact) vim.ui.img.ConvertOpts
+---@field background? string hex string representing background color
+---@field crop? vim.ui.img.utils.Region
+---@field format? string such as 'png', 'rgb', or 'sixel' (default 'png')
+---@field out? string write to output instead of stdout
+---@field quality? integer default for PNG is 75 (see https://imagemagick.org/script/command-line-options.php#quality)
+---@field size? vim.ui.img.utils.Size
+---@field timeout? integer maximum time (in milliseconds) to wait for conversion
+
+---Converts an image using ImageMagick, returning the bytes of the new image.
+---
+---If `background` is specified, will convert alpha pixels to the background color (e.g. #ABCDEF).
+---If `crop` is specified, will crop the image to the specified pixel dimensions.
+---If `format` is specified, will convert to the image format, defaulting to png.
+---If `out` is specified, will treat as output file and write to it, and promise returns empty str.
+---If `size` is specified, will resize the image to the desired size.
+---@param opts? vim.ui.img.ConvertOpts
+---@return vim.ui.img.utils.Promise<string>
+function M:convert(opts)
```

Comment:
out of scope

Reply:
Not a problem. If we did want `sixel` or need cropping for `iterm2` (if we brought it in), then we would need this. Or we'd have to provide or write a PNG decoder.

Reply:
requiring PNG seems reasonable. converting to PNG is a different topic, which could be covered in some other api in the future. but the caller of this api is expected to provide PNG.

Reply:
Yes, I think supporting png only is reasonable. It also makes storing the images in memory more efficient, which we will need to do, since Kitty has a limited buffer size and images can be evicted.

What makes the evection even worse, is that on tmux with pass through, we don't have a way to know when that happens or how much quota is used. So, I think the only safe way to do it, is to re-upload the image to kitty each time a new image is brought to view.

Reply:
@fredizzimo in snacks they use a least recently used queue to know which images to re-upload. I was going to implement that, but held off when discussion started back.

Or we transmit again every time, but that definitely has noticeable performance considerations from what I saw with iterm2 and sixel.

LRU out of scope for this PR? Retransmitting out of scope for this PR?

Reply:
I think we only need to retransmit when a new image appears, not for every movement, but yes it could still be annoying when scrolling up and down.

LRU, with some size limit could work. The only thing we really have to be careful with is when the image is displayed in tmux or similar, in the other cases, neovim has full control.  And the same thing applies to unique id collisions, not much we can do about that in a tmux context.

Since this does not affect the interface, I would vote for out of scope.

Reply:
For LRU we have `vim.ringbuf`. But yes, let's save performance improvements for a followup, not here.